lock-impl/src/main/java/com/palantir/lock/impl/RateLimitingLockService.java

```diff
@@ -0,0 +1,185 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.lock.impl;
+
+import java.math.BigInteger;
+import java.util.Set;
+
+import javax.annotation.Nullable;
+
+import com.palantir.lock.HeldLocksGrant;
+import com.palantir.lock.HeldLocksToken;
+import com.palantir.lock.LockClient;
+import com.palantir.lock.LockRefreshToken;
+import com.palantir.lock.LockRequest;
+import com.palantir.lock.LockResponse;
+import com.palantir.lock.LockServerOptions;
+import com.palantir.lock.LockService;
+import com.palantir.lock.SimpleHeldLocksToken;
+
+public class RateLimitingLockService implements LockService{
+    private static final ThreadCountLimiter globalLimiter = new ThreadCountLimiter(0);
+
+    private LockService delegate;
+    private ThreadCountLimiter limiter;
+
+    private RateLimitingLockService() {
+        // hidden
+    }
+
+    public static RateLimitingLockService create(LockService delegate, int maxServerThreads, int numClients) {
+        int threadPoolSize = maxServerThreads - 15;
+        int localLimiterSize = threadPoolSize / numClients / 2;
+        int globalLimiterSize = threadPoolSize - localLimiterSize * numClients;
+
+        RateLimitingLockService service = new RateLimitingLockService();
+        service.delegate = delegate;
+        service.limiter = new ThreadCountLimiter(localLimiterSize);
+        globalLimiter.set(globalLimiterSize);
+        return service;
+    }
+
+
+    @Nullable
+    @Override
+    public LockRefreshToken lock(String client, LockRequest request) throws InterruptedException {
+        if (limiter.acquire()) {
+            LockRefreshToken token = delegate.lock(client, request);
+            limiter.release();
+            return token;
+        } else if (globalLimiter.acquire()) {
+            LockRefreshToken token = delegate.lock(client, request);
+            globalLimiter.release();
+            return token;
+        } else {
+            throw new TooManyRequestsException();
+        }
+    }
+
+    @Override
+    public HeldLocksToken lockAndGetHeldLocks(String client, LockRequest request) throws InterruptedException {
+        return null;
+    }
+
+    @Override
+    public boolean unlock(LockRefreshToken token) {
+        return delegate.unlock(token);
+    }
+
+    @Override
+    public Set<LockRefreshToken> refreshLockRefreshTokens(Iterable<LockRefreshToken> tokens) {
+        if (limiter.acquire()) {
+            Set<LockRefreshToken> result = delegate.refreshLockRefreshTokens(tokens);
+            limiter.release();
+            return result;
+        } else if (globalLimiter.acquire()) {
+            Set<LockRefreshToken> result = delegate.refreshLockRefreshTokens(tokens);
+            globalLimiter.release();
+            return result;
+        } else {
+            throw new TooManyRequestsException();
```

Comment:
might want to extract a method that performs an operation with local limiter, global limiter, or throws, and makes sure to release any acquired limiter